vllm/worker/hpu_model_runner.py

```diff
@@ -278,17 +278,25 @@ def flatten(in_list):
     return list(itertools.chain(*in_list))
 
 
-def modify_decoder_layer(module: torch.nn.Module, suffix="DecoderLayer"):
-    if module.__class__.__name__.endswith(suffix):
+def modify_decoder_layer(module: torch.nn.Module,
+                         suffix="DecoderLayer",
+                         n=1,
+                         counter=None):
 
-        def forward_hook(module, args, output):
-            htorch.core.mark_step()
-            return output
+    def forward_hook(module, args, output):
+        htorch.core.mark_step()
+        return output
 
-        module.register_forward_hook(forward_hook)
+    if counter is None:
+        counter = [0]
 
     for child_name, child_module in module.named_children():
```

Comment:
Is the call to register_forward_hook on 'module' omitted on purpose? If n=1 mark_step will be called first after forward of first child module. Previously it was called first after forward of top module also.

Reply:
Umm not sure if I understand your question correctly.  The behavior of n=1 is exactly same from previous code and this new code. Even previous code, only when the module matches with suffix(DecoderLayer), then only register the forward hook.  I had to move the register_forward_hook on module inside the for loop so I can keep the counter. 

**Previous code:**  => For llama model 32x Decoderlayer, mark_step will be called 32 times for every decoding step.
```
=Check model : LlamaModel
=Check embed_tokens : VocabParallelEmbedding
=Check layers : ModuleList
=Check 0 : LlamaDecoderLayer
===Register HOOK: LlamaDecoderLayer
=Check self_attn : LlamaAttention
...
...
=Check post_attention_layernorm : RMSNorm
=Check 1 : LlamaDecoderLayer
===Register HOOK: LlamaDecoderLayer
=Check self_attn : LlamaAttention
...
```

**New code (n=1)** => same, mark_step will be called 32 times for every decoding step.
```
=Check model : LlamaModel
=Check embed_tokens : VocabParallelEmbedding
=Check layers : ModuleList
=Check 0 : LlamaDecoderLayer
===Register HOOK: LlamaDecoderLayer
=Check 1 : LlamaDecoderLayer
===Register HOOK: LlamaDecoderLayer
=Check 2 : LlamaDecoderLayer
===Register HOOK: LlamaDecoderLayer
=Check 3 : LlamaDecoderLayer
===Register HOOK: LlamaDecoderLayer
=Check 4 : LlamaDecoderLayer
===Register HOOK: LlamaDecoderLayer
=Check 5 : LlamaDecoderLayer
...
=Check 31 : LlamaDecoderLayer
===Register HOOK: LlamaDecoderLayer
=Check norm : RMSNorm
=Check lm_head : ParallelLMHead
=Check logits_processor : LogitsProcessor
=Check sampler : Sampler
```

**New code (n=8)** => mark_step will be called 4(32/8) times for every decoding step.
```
=Check model : LlamaModel
=Check embed_tokens : VocabParallelEmbedding
=Check layers : ModuleList
=Check 0 : LlamaDecoderLayer
=Check 1 : LlamaDecoderLayer
=Check 2 : LlamaDecoderLayer
=Check 3 : LlamaDecoderLayer
=Check 4 : LlamaDecoderLayer
=Check 5 : LlamaDecoderLayer
=Check 6 : LlamaDecoderLayer
=Check 7 : LlamaDecoderLayer
===Register HOOK: LlamaDecoderLayer
=Check 8 : LlamaDecoderLayer
=Check 9 : LlamaDecoderLayer
=Check 10 : LlamaDecoderLayer
=Check 11 : LlamaDecoderLayer
=Check 12 : LlamaDecoderLayer
=Check 13 : LlamaDecoderLayer
=Check 14 : LlamaDecoderLayer
=Check 15 : LlamaDecoderLayer
===Register HOOK: LlamaDecoderLayer
```